File: src/main/java/com/jore/epoc/bo/ProductLifecycle.java
package com.jore.epoc.bo;

import com.jore.Assert;
import com.jore.util.NormalDistribution;

public class ProductLifecycle {
    private final static NormalDistribution NORMAL_DISTRIBUTION = new NormalDistribution();
    private final static double STANDARD_DEVIATION_SPAN = 4.6d;
    private final static double TAIL_ERROR = 0.01072411002d;
    private final double productLifecycleDuration;

    public ProductLifecycle(double d) {
        this.productLifecycleDuration = d;
    }

    public double getPercentageSoldForMonths(int months) {
        Assert.isTrue("Duration not initialized", productLifecycleDuration > 0);
        final double durationFraction = months / productLifecycleDuration;
        final double standardDeviationFraction = -(STANDARD_DEVIATION_SPAN / 2) + (durationFraction * STANDARD_DEVIATION_SPAN);
        double percentageSold = NORMAL_DISTRIBUTION.distributionFor(standardDeviationFraction) - TAIL_ERROR;
        percentageSold += (percentageSold * TAIL_ERROR * 2);
        return Math.max(Math.min(percentageSold, 100), 0);
    }
}

File: EpocServer/Sources/App/Models/ProductLifecycle.swift
import Foundation

/// Models the S-curve product adoption over time using a normal distribution.
/// Equivalent to `com.jore.epoc.bo.ProductLifecycle`.
///
/// A new product reaches full market penetration gradually. This class returns
/// the percentage of the market reachable at a given number of months since launch.
public struct ProductLifecycle: Sendable {
    private static let normalDistribution = NormalDistribution()
    private static let standardDeviationSpan = 4.6
    private static let tailError = 0.01072411002

    private let productLifecycleDuration: Double

    public init(_ duration: Double) {
        self.productLifecycleDuration = duration
    }

    /// Returns the percentage of the market sold/reachable after the given number of months.
    /// Returns a value between 0 and ~1.0 (100%).
    public func getPercentageSoldForMonths(_ months: Int) -> Double {
        precondition(productLifecycleDuration > 0, "Duration not initialized")
        let durationFraction = Double(months) / productLifecycleDuration
        let standardDeviationFraction = -(Self.standardDeviationSpan / 2) + (durationFraction * Self.standardDeviationSpan)
        var percentageSold = Self.normalDistribution.distributionFor(standardDeviationFraction) - Self.tailError
        percentageSold += (percentageSold * Self.tailError * 2)
        return max(min(percentageSold, 100), 0)
    }
}
